node/grpc/v2_metrics.go

```diff
@@ -0,0 +1,175 @@
+		prometheus.SummaryOpts{
+			Namespace:  namespace,
+			Name:       "store_chunks_latency_ms",
+			Help:       "The latency of a StoreChunks() RPC call.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{},
+	)
+
+	storeChunksDataSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "store_chunks_data_size_bytes",
+			Help:      "The size of the data requested to be stored by StoreChunks() RPC calls.",
+		},
+		[]string{},
+	)
+
+	getChunksLatency := promauto.With(registry).NewSummaryVec(
+		prometheus.SummaryOpts{
+			Namespace:  namespace,
+			Name:       "get_chunks_latency_ms",
+			Help:       "The latency of a GetChunks() RPC call.",
+			Objectives: map[float64]float64{0.5: 0.05, 0.9: 0.01, 0.99: 0.001},
+		},
+		[]string{},
+	)
+
+	getChunksDataSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "get_chunks_data_size_bytes",
+			Help:      "The size of the data requested to be retrieved by GetChunks() RPC calls.",
+		},
+		[]string{},
+	)
+
+	dbSize := promauto.With(registry).NewGaugeVec(
+		prometheus.GaugeOpts{
+			Namespace: namespace,
+			Name:      "db_size_bytes",
+			Help:      "The size of the leveldb database.",
+		},
+		[]string{},
+	)
+	isAlive := &atomic.Bool{}
+	isAlive.Store(true)
+
+	return &V2Metrics{
+		logger:              logger,
+		registry:            registry,
+		grpcServerOption:    grpcServerOption,
+		storeChunksLatency:  storeChunksLatency,
+		storeChunksDataSize: storeChunksDataSize,
+		getChunksLatency:    getChunksLatency,
+		getChunksDataSize:   getChunksDataSize,
+		dbSize:              dbSize,
+		dbSizePollPeriod:    dbSizePollPeriod,
+		dbDir:               dbDir,
+		isAlive:             isAlive,
+	}, nil
+}
+
+// Start starts the metrics server.
+func (m *V2Metrics) Start() {
+	if m.dbSizePollPeriod.Nanoseconds() == 0 {
+		return
+	}
+	go func() {
+		ticker := time.NewTicker(m.dbSizePollPeriod)
+
+		for m.isAlive.Load() {
+			var size int64
+			err := filepath.Walk(m.dbDir, func(_ string, info os.FileInfo, err error) error {
+				if err != nil {
+					return err
+				}
+				if !info.IsDir() {
+					size += info.Size()
+				}
+				return err
+			})
+
+			if err != nil {
+				m.logger.Errorf("failed to get database size: %v", err)
+			} else {
+				m.dbSize.WithLabelValues().Set(float64(size))
+			}
+			<-ticker.C
+		}
+	}()
+
+}
+
+// Stop stops the metrics server.
+func (m *V2Metrics) Stop() {
+	m.isAlive.Store(false)
+}
+
+// GetGRPCServerOption returns the gRPC server option that enables automatic GRPC metrics collection.
+func (m *V2Metrics) GetGRPCServerOption() grpc.ServerOption {
+	return m.grpcServerOption
+}
+
+func (m *V2Metrics) ReportStoreChunksLatency(latency time.Duration) {
+	m.storeChunksLatency.WithLabelValues().Observe(
+		float64(latency.Nanoseconds()) / float64(time.Millisecond))
```

Comment:
float64(latency.Milliseconds()) should work

Reply:
I'm intentionally not using that (@ian-shim made the same suggestion on another PR 😜). `latency.Milliseconds()` returns an `int`, meaning we lose all sub-millisecond fidelity in the measurement. For many metrics this isn't a hug deal (e.g. if you are measuring something that takes 100s of milliseconds), but for some the precision is nice to have.

Let me know if you'd like to discuss this further.

Reply:
I think it depends on what the expected number we will deal with here. If it's ultra low latency and sub-ms matters, `Microseconds()` may be an option.

Reply:
For the sake of consistency, I've been converting everything to `ms` for reporting latencies.

Although I can often make an educated guess as to the expected latency for an operation, absent experimental data it's only a guess. If I guess wrong, then we could end up in a situation where we guess wrong.

Would this be a topic you think worthwhile to schedule a short call to discuss?

Reply:
As discussed, all of these now use a utility method in common.